src/client/game.cpp

```diff
@@ -3266,144 +3251,162 @@ bool Game::nodePlacementPrediction(const ItemDefinition &selected_def,
 	bool is_valid_position;
 
 	node = map.getNode(nodepos, &is_valid_position);
-	if (!is_valid_position)
+	if (!is_valid_position) {
+		soundmaker->m_player_rightpunch_sound = selected_def.sound_place_failed;
 		return false;
+	}
 
-	if (!prediction.empty() && !(nodedef->get(node).rightclickable &&
+	if (prediction.empty() || (nodedef->get(node).rightclickable &&
 			!isKeyDown(KeyType::SNEAK))) {
-		verbosestream << "Node placement prediction for "
-			<< selected_item.name << " is "
-			<< prediction << std::endl;
-		v3s16 p = neighbourpos;
-
-		// Place inside node itself if buildable_to
-		MapNode n_under = map.getNode(nodepos, &is_valid_position);
-		if (is_valid_position)
-		{
-			if (nodedef->get(n_under).buildable_to)
-				p = nodepos;
-			else {
-				node = map.getNode(p, &is_valid_position);
-				if (is_valid_position &&!nodedef->get(node).buildable_to)
-					return false;
+		// Report to server
+		client->interact(INTERACT_PLACE, pointed);
```

Comment:
since it needs to do this anyway, why is this not a the top-level of the function?

Reply:
It isn't done anyway, see https://github.com/minetest/minetest/pull/8640/files#diff-7f71a92de20504e8a5636b4317ea67b2L3270, https://github.com/minetest/minetest/pull/8640/files#diff-7f71a92de20504e8a5636b4317ea67b2R3400 and https://github.com/minetest/minetest/pull/8640/files#diff-7f71a92de20504e8a5636b4317ea67b2R3408.

Reply:
Hmm ok, I'm not 100% happy with the code but it should be fine.